File: _Programmers-고득점-Kit/08_깊이 너비 우선 탐색(DFS BFS)/PG-Lv2_게임 맵 최단거리.java
import java.util.*;

class Solution {
    int bfs(int n, int m, int[][] maps) {
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        int[][] distance = new int[n][m];
        
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0});
        maps[0][0] = 0; // visited
        
        while (!q.isEmpty()) {
            int[] pos = q.poll();
            int r = pos[0];
            int c = pos[1];
            
            // 도달했다면 return
            if (r == n - 1 && c == m - 1) {
                return distance[r][c] + 1;
            }
            
            // 상하좌우 진행
            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < m && maps[nr][nc] == 1) {
                    q.offer(new int[]{nr, nc});
                    maps[nr][nc] = 0; // visited
                    distance[nr][nc] = distance[r][c] + 1; // distance
                }
            }
        }
        
        // 도달하지 못했다면 -1 return
        return -1;
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        int n = maps.length;
        int m = maps[0].length;
        
        answer = bfs(n, m, maps);
        
        return answer;
    }
}

File: _Programmers-고득점-Kit/08_깊이 너비 우선 탐색(DFS BFS)/PG-Lv2_게임 맵 최단거리.py
# [PG] 1844 - 게임 맵 최단거리 (Lv2)
# https://school.programmers.co.kr/learn/courses/30/lessons/1844

from collections import deque


def solution(maps):
    """
    최단거리를 찾는다   -> bfs로 접근하면서, visit 하지 않은 곳만을 방문한다.
                    -> 이전 위치의 값에서 1씩 늘려가며 저장한다.
    """
    n, m = len(maps), len(maps[0])

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    q = deque([(0, 0)])

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            # 범위 확인 & 벽인지 확인
            if nx < 0 or nx >= n or ny < 0 or ny >= m or not maps[nx][ny]:
                continue

            # 방문하지 않았던 곳인지 확인
            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                q.append((nx, ny))

    return maps[-1][-1] if maps[-1][-1] > 1 else -1


def solution2(maps):
    answer = 0

    """
    BFS (최단거리)
    - 목표에 도달하면 거리
    - 목표에 도달하지 못하면 -1
    """

    dr, dc = [-1, 1, 0, 0], [0, 0, -1, 1]
    N, M = len(maps), len(maps[0])

    def bfs():
        q = deque([(0, 0)])
        distance = [[0] * M for _ in range(N)]  # 거리
        maps[0][0] = 0  # visited 처리

        while q:
            r, c = q.popleft()

            if r == N - 1 and c == M - 1:
                return distance[r][c] + 1

            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if 0 <= nr < N and 0 <= nc < M and maps[nr][nc]:
                    q.append((nr, nc))
                    maps[nr][nc] = 0
                    distance[nr][nc] = distance[r][c] + 1

        return -1

    answer = bfs()

    return answer


maps = [[1,0,1,1,1],[1,0,1,0,1],[1,0,1,1,1],[1,1,1,0,1],[0,0,0,0,1]]
assert 11 == solution(maps)
